### PR TITLE
Add Faults and Fault Suppression

### PR DESCRIPTION
**Pull request by Duncan-McD**
# Add Faults and Fault Suppression

Fixes [FS-180](https://ssdsalphacubesat.atlassian.net/browse/FS-180?atlOrigin=eyJpIjoiNWY5M2NmYTI4YjZmNDI5OWJhMDJlNTA2YmMzNDZhZDkiLCJwIjoiaiJ9)

### Summary of changes
- Add Faults
- Refactor current faults across software
- Tests for Faults
- Revised normal report to downlink all the fault information

### Testing
I added basic test cases to test the faults. These test cases (test_faults) is passing in the teensy35 environment.

I also tested the Fault commands using the rockblock simulator and was successfully able to send commands to suppress and force a fault for an imu mag_x fault.

I then switched to the flatsat testbed and tested the same series of commands with rockblock in the loop. I sent commands from the rockblock portal, and then checked the corresponding normal reports to verify the data sent down in the reports make sense.

Byte 45 of the Normal report was the byte corresponding to fault_groups::imu_faults::mag_x  and I was successfully able to see a transition of this field from:

9 -> D -> 3 -> 9

The 9 cooresponds to: 1001 (Or Base and signaled high. This is beacause the IMU was invalid on the flatsat prior to sending any commands)

Then I sent the command for Forcing this fault (6000 0000 0001 0000 0000 00FA) and the next downlink showed D (1101) Which has Base, Signaled, and Forced high.

Then I sent the command for Suppressing this fault (6000 0000 0000 0000 0001 00FA) and the next downlink showed 3 (0011) Which is only suppressed and signaled high.

Finally I sent the command for restoring signal to the fault (6000 0000 0000 0000 0000 00FA) and the next downlink showed 9 once again (With base and signaled high).

### SFR Changes
Each SensorReading in the SFR was changed to intake a Fault pointer instead.

### Documentation Evidence
Post to the wiki, or one of the following
- I added extensive docs in the code to describe all of the functions related to my code.

[FS-180]: https://ssdsalphacubesat.atlassian.net/browse/FS-180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: platformio.ini [hunk 1]
```diff
@@ -36,6 +36,7 @@ platform = teensy
 board = teensy35
 framework = arduino
 upload_protocol = teensy-cli
+test_build_src = true
```
Comment (ljgreenhill) [inline platformio.ini:39]: should be removed- deprecated

File: src/SensorReading.cpp [hunk 2]
```diff
@@ -93,34 +52,16 @@ void SensorReading::set_value(float x)
...
-            faults::fault_2 &= ~map_to_mask[this->type];
-        } else {
-            faults::fault_3 &= ~map_to_mask[this->type];
-        } // clear the flag in the corresponding fault register
+    if (fault != NULL) {
+        fault->release();
     }
 }
 
 void SensorReading::set_invalid()
```
Comment (ljgreenhill) [inline src/SensorReading.cpp:60]: where are buffers cleared if invalidated?

File: platformio.ini (hunk 1 above)
Comment (Duncan-McD) [reply platformio.ini:39]: This setting just enables building src code when invoking a "pio test" and doesn't affect pio build or pio run - not including this setting means any pio test in the teensy35 environment will fail

Is there a particular reason it was removed?

File: src/SensorReading.cpp (hunk 2 above)
Comment (Duncan-McD) [reply src/SensorReading.cpp:60]: Oh that definitely wasn't supposed to be deleted - good catch

Comment (Duncan-McD) [reply src/SensorReading.cpp:60]: changed this to If a fault was initialized and we are not suppressed, clear buffer and if a fault was not initialized always clear buffer

**Pull request merged**